linalg/constraints.hpp

```diff
@@ -0,0 +1,473 @@
+};
+
+
+/** Uses CG and a HypreBoomerAMG preconditioner for the penalized system. */
+class PenaltyPCGSolver : public PenaltyConstrainedSolver
+{
+public:
+   PenaltyPCGSolver(HypreParMatrix& A, SparseMatrix& B, double penalty_,
+                    int dimension=0, bool reorder=false) :
+      PenaltyConstrainedSolver(A, B, penalty_),
+      dimension_(dimension), reorder_(reorder)
+   { }
+
+   PenaltyPCGSolver(HypreParMatrix& A, HypreParMatrix& B, double penalty_,
+                    int dimension=0, bool reorder=false) :
+      PenaltyConstrainedSolver(A, B, penalty_),
+      dimension_(dimension), reorder_(reorder)
+   { }
+
+protected:
+   virtual Solver* BuildPreconditioner() const override
+   {
+      HypreBoomerAMG* h_prec = new HypreBoomerAMG(*penalized_mat);
+      h_prec->SetPrintLevel(0);
+      if (dimension_ > 0) { h_prec->SetSystemsOptions(dimension_, reorder_); }
+      return h_prec;
+   }
+
+   virtual IterativeSolver* BuildKrylov() const override
+   { return new CGSolver(GetComm()); }
+
+private:
+   int dimension_;
+   bool reorder_;
+};
+
+#endif
+
+/** @brief Solve constrained system by solving original mixed sysetm;
+    see ConstrainedSolver.
+
+    Solves the saddle-point problem with a block-diagonal preconditioner, with
+    user-provided preconditioner in the top-left block and (by default) an
+    identity matrix in the bottom-right.
+
+    This is the most general ConstrainedSolver, needing only Operator objects
+    to function. But in general it is not very efficient or scalable. */
+class SchurConstrainedSolver : public ConstrainedSolver
+{
+public:
+   /// Setup constrained system, with primal_pc a user-provided preconditioner
+   /// for the top-left block.
+#ifdef MFEM_USE_MPI
+   SchurConstrainedSolver(MPI_Comm comm, Operator& A_, Operator& B_,
+                          Solver& primal_pc_);
+#endif
+   SchurConstrainedSolver(Operator& A_, Operator& B_, Solver& primal_pc_);
+   virtual ~SchurConstrainedSolver();
+
+   virtual void Mult(const Vector& x, Vector& y) const override;
+
+protected:
+#ifdef MFEM_USE_MPI
+   SchurConstrainedSolver(MPI_Comm comm, Operator& A_, Operator& B_);
+#endif
+   SchurConstrainedSolver(Operator& A_, Operator& B_);
+
+   Array<int> offsets;
+   BlockOperator * block_op;  // owned
+   TransposeOperator * tr_B;  // owned
+   Solver * primal_pc; // NOT owned
+   BlockDiagonalPreconditioner * block_pc;  // owned
+   Solver * dual_pc;  // owned
+
+private:
+   void Initialize();
+};
+
+
+#ifdef MFEM_USE_MPI
+/** @brief Basic saddle-point solver with assembled blocks (ie, the
+    operators are assembled HypreParMatrix objects.)
+
+    This uses a block-diagonal preconditioner that approximates
+    \f$ [ A^{-1} 0; 0 (B A^{-1} B^T)^{-1} ] \f$.
+
+    In the top-left block, we approximate \f$ A^{-1} \f$ with HypreBoomerAMG.
+    In the bottom-right, we approximate \f$ A^{-1} \f$ with the inverse of the
+    diagonal of \f$ A \f$, assemble \f$ B D^{-1} B^T \f$, and use
```

Comment:
D^{-1} = diag(A)^{-1}? If yes, could it be beneficial to do an l1 diagonal inverse?

Reply:
Maybe? This is a very easy change to make if you want me to, but I haven't tried it.

Reply:
It just came to my mind in the context of if its useful expose a custom preconditioning operator